File: design_patterns/6-builder/descontos.py
class Desconto_por_cinco_items(object):
    def __init__(self, proximo_desconto):
        self.__proximo_desconto = proximo_desconto

    def calcula(self, orcamento):
        if orcamento.total_items > 5:
            return round(orcamento.valor * 0.1, 2)
        else:
            return self.__proximo_desconto.calcula(orcamento)


class Desconto_por_mais_de_quinhentos_reais(object):
    def __init__(self, proximo_desconto):
        self.__proximo_desconto = proximo_desconto

    def calcula(self, orcamento):
        if orcamento.valor > 500:
            return round(orcamento.valor * 0.07, 2)
        else:
            return self.__proximo_desconto.calcula(orcamento)


class Sem_desconto(object):
    def calcula(self):
        return 0

File: design_patterns/7-observer/descontos.rb
class DescontoPorQuantidade
  def initialize(proximo_desconto)
    @proximo_desconto = proximo_desconto
  end

  def calcula(orcamento)
    puts "Quantidade #{orcamento.qtd_items}"
    if orcamento.qtd_items > 5
      puts 'Aplicando Desconto pela Quantidade'
      (orcamento.total * 0.1).round(2)
    else
      @proximo_desconto.calcula(orcamento)
    end
  end
end

class DescontoPorValor
  def initialize(proximo_desconto)
    @proximo_desconto = proximo_desconto
  end

  def calcula(orcamento)
    puts "Total do Orcamento #{orcamento.total}"
    if orcamento.total > 500
      puts 'Aplicando Desconto pelo Valor'
      (orcamento.total * 0.07).round(2)
    else
      @proximo_desconto.calcula(orcamento)
    end
  end
end

class SemDesconto
  def calcula(_obj)
    0
  end
end
